RescoreBuilder: Add parsing and creating of RescoreSearchContext

cbuescher opened the pull request:
Adding the ability to parse from xContent to the rescore builder. Also making RescoreBuilder an abstract base class that encapsulates the `window_size` setting, with QueryRescoreBuilder as its only implementation at the moment.

Relates to #15559

cbuescher commented:
I added a second part to this PR which switches from using BytesReference to the refactored RescoreBuilder in SearchSourceBuilder.

colings86 commented on core/src/main/java/org/elasticsearch/search/rescore/RescoreBaseBuilder.java:
```diff
@@ -0,0 +1,175 @@
+ * The base builder for rescorers. Wraps a conrete instance of {@link RescoreBuilder} and
+ * adds the ability to specify the optional `window_size` parameter
+ */
+public class RescoreBaseBuilder implements ToXContent, Writeable<RescoreBaseBuilder> {
```
Given that all rescorers must have a window size set and we have got rid of the default window size setting so they must be explicitly set on the rescorer itself before being passed to the search request I think we should not have this class and instead have window size set on the RescoreBuilder so that all implementations of RescoreBuilder have the windowSize implicitly. This will also simplify both parsing form XContent and the Java API.

cbuescher replied:
Thats a good point, I will look into this. Will require some small changes to the Java-API but I would also prefer to get rid of this class.

colings86 commented on core/src/main/java/org/elasticsearch/search/rescore/RescoreParseElement.java:
```diff
@@ -71,7 +74,7 @@ public void parseSingleRescoreContext(XContentParser parser, SearchContext conte
             rescoreContext.setWindowSize(windowSize.intValue());
         }
-        context.addRescore(rescoreContext);
+        return rescoreContext;
```
Why does this class need to change? the workings of this class should not need to change since it is only used on the shards and will be removed when the search refactoring is completed. I think we should leave it as it was so we have a reference to how the functionality used to be in case we need to refer to it.

cbuescher replied:
This enables us to do roundtrip-tests for the new `build()` method as long as this class is not gone. It's the same approach I took with the highlighters. The utility method `parseSingleRescoreContext` now returns a RescoreSearchContext rather than requiring a SearchContext to set it on. I think this change is small enough given that it enables us to check that we build the same RescoreSearchContext objects like before in the tests.

colings86 commented:
@cbuescher I left a few comments

cbuescher commented:
@colings86 thanks, I added a commit that changes the base builder holding the window_size to be an abstract superclass of QueryRescoreBuilder, also making RescoreBuilder the common interface. Although there is only one implementation at the moment this should enable possible future extensions.

colings86 commented on core/src/main/java/org/elasticsearch/search/rescore/RescoreBuilder.java:
```diff
@@ -19,256 +19,20 @@
-import java.util.Objects;
 
-public class RescoreBuilder implements ToXContent, Writeable<RescoreBuilder> {
+public interface RescoreBuilder<RB extends RescoreBuilder<?>> extends ToXContent, NamedWriteable<RB> {
```
Do we need both the interface and the abstract class?

cbuescher replied:
I don't think we need it, but thats the way we also do it with QueryBuilders (Interface, Abstract base class). I wonder if we should change the name of the abstract class to RescoreBuilder then, that would be similar to HighlightBuilder and ShapeBuilders then, wdyt?

colings86 replied:
+1 to removing the interface and ranaming the abstract class to RescoreBuilder

colings86 commented:
left one more comment but LGTM

cbuescher commented:
@colings86 thanks for the review, I renamed the abstract base class to RescoreBuilder as suggested and added a note about the merge of the former RescoreBuilder with the inner RescoreBuilder.Rescorer class to the migration doc.
